fog/view/server/src/server.rs

```diff
@@ -397,21 +403,36 @@ where
 
             // Insert the records into the enclave.
             self.add_records_to_enclave(
-                fetched_records.ingest_invocation_id,
+                fetched_records.ingress_key,
                 fetched_records.block_index,
                 fetched_records.records,
             );
         }
 
         // Figure out the higehst fully processed block count and put that in the shared state.
-        let ingestable_ranges = self.db_fetcher.known_ingestable_ranges();
-        let missing_block_ranges = self.db_fetcher.known_missing_block_ranges();
+        let (highest_known_block_index, ingress_keys, missing_block_ranges) =
+            self.db_fetcher.get_highest_processed_block_context();
+        let (highest_processed_block_count, reason_we_stopped) = self
+            .enclave_block_tracker
+            .highest_fully_processed_block_count(
+                highest_known_block_index,
+                &ingress_keys,
+                &missing_block_ranges,
+            );
 
         let mut shared_state = self.shared_state.lock().expect("mutex poisoned");
-        let highest_processed_block_count = self
-            .enclave_block_tracker
-            .highest_fully_processed_block_count(&ingestable_ranges, &missing_block_ranges);
-        shared_state.highest_processed_block_count = highest_processed_block_count;
+        if shared_state.highest_processed_block_count != highest_processed_block_count {
+            shared_state.highest_processed_block_count = highest_processed_block_count;
+            self.last_unblocked_at = Instant::now();
+        } else if self.last_unblocked_at.elapsed() >= Duration::from_secs(60) {
+            if let Some(reason_we_stopped) = reason_we_stopped {
+                log::warn!(self.logger, "We seem to be stuck at highest_processed_block_count = {} for at least a minute... we are blocked on an ingress key making progress: {:?}", highest_processed_block_count, reason_we_stopped);
+            } else {
+                log::info!(self.logger, "We seem to be stuck at highest_processed_block_count = {} for at least a minute... we have processed all blocks known to the recovery database", highest_processed_block_count);
```

Comment:
Wouldn't this log once a minute for as long as there is no activity in the network?

Reply:
yeah, is that bad? i guess we could just remove this line

Reply:
I am not sure. Its nice to see things are still alive...
@jgreat @joekottke any thoughts on how much noise is too much noise? Context here is a log message that happens once a minute when the network is inactive. The log message lets us know fog-view is still alive but doesn't think it has anything to work on.

Reply:
@eranrund, depends on your budget for datadog. Ideally I only want things I should actually react to as a normal engineer. Example: Give one alert for when network isnt connecting as an error log, but info logs should be responding with this heartbeat when its not working, this way I can debug better when things are wrong, but dont have to pay 10000/mo for the 10m log lines this could leave on datadog..

Reply:
If this isn't an error condition, I'd make this a debug message. Useful logging is usually about events in the system.

This kind of information we should be able to collect via metrics. If we see an abnormal pattern in blocks processes or a metric that shows which block we are on that we can compare with the current block on the chain, we can alert off of that.

Reply:
so here's the issue i see there:

1) We can alert based on highest_processed_block_count being behind the block chain, by making a prometheus counter
2) The piece of information you need then is, the hex of the ingress key that we are stuck on
3) Then we go look at the ingest servers and see if they are all dead or one of them just needs to be activated
4) If they are all dead then we report a lost key to the database (literally that hex string)

So you really kind of want this "reason_we_stopped" value
I don't see how I can stick that value in a prometheus metric, or maybe I just don't know enough about prometheus? it seems like a logs / kibana thing

my other idea was, we could make a grpc API so that the view server can just be asked what key it is stuck on, and that's how you get the info rather than the logs?

lmk what you think

Reply:
i have demoted this to debug for now, i'm going to make a separate JIRA ticket to expose this somehow

Reply:
this is now fog-425